### PR TITLE
[E2E] Improve + Fix tests that use sandboxing

### PR DESCRIPTION
**Pull request by sloansparger**
Pursuant of #47095

### Description

Improvements based on these internal slack threads: https://metaboat.slack.com/archives/C505ZNNH4/p1722611219985539
https://metaboat.slack.com/archives/C505ZNNH4/p1724187422396359

**Problem**
Permissions for some sandbox e2e tests were misconfigured such that they had full view access resulting in false positives. Many of our sandbox tests were not making any assertions on wether they were sandboxed, but only on the related functionality we wanted to test.

**Solution**
Add 3 layers of assertions on tests that test on sandboxed data where possible:
1. Assert the results are limited (and I added comments with # of results all access would have)
2. Assert on data request having `is_sandboxed` set to `true` in the response
3. Assert on the data in the response following some expected constraints (typically that all values in a column equal the user attribute value we're sandboxing on)

The last two points make use of a new util `assertDatasetReqIsSandboxed` (its api is documented in the code). In some cases the data that comes back is aggregated so we're unable to make the third assertion. In this case we'll just fall back to `is_sandboxed` being set to `true`.

**Note**
After combing through these tests we had a few other sandboxes that were misconfigured. Those have been fixed as well.

### How to verify

- You should be able to replace  `cy.signInAsSandboxedUser` with `cy.signInAsANormalUser` and see failures for almost all sandbox tests. Note: Some test for the absence of behavior which is consistent with unrestricted access or the test is configuring sandboxes.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: e2e/test/scenarios/permissions/sandboxes.cy.spec.js [hunk 1]
```diff
@@ -221,6 +215,13 @@ describeEE("formatting > sandboxes", () => {
         },
       });
 
+      cy.sandboxTable({
+        table_id: PEOPLE_ID,
+        attribute_remappings: {
+          attr_uid: ["dimension", ["field", PEOPLE.ID, null]],
+        },
+      });
+
```
Comment (nemanjaglumac) [inline e2e/test/scenarios/permissions/sandboxes.cy.spec.js:224]: This wasn't so much about the order, as much as it was about granting the "unrestricted" `view-data` permissions on a database level after sandboxing a table (basically making the previous action void).

This is how the test looked before the "split permissions"
https://github.com/metabase/metabase/blob/42ce333d5140ddd7381bba7fef8caefeb28a3f64/e2e/test/scenarios/permissions/sandboxes.cy.spec.js#L199-L214


1. We were staring with both "All users" and "Collection" groups data permissions set to "No self-service"
2. We then sandboxed the People table by matching the user attribute to the People.id
3. We then lifted the data restrictions for the three remaining tables (orders, products, reviews) but we didn't break sandboxing - this was targeted and selective

The test rewrite broke the sandboxing.
The correct payload should've been:
```js
cy.updatePermissionsGraph({
  [COLLECTION_GROUP]: {
    [SAMPLE_DB_ID]: {
      "view-data": {
        "PUBLIC": {
          [ORDERS_ID]: "unrestricted",
          [PRODUCTS_ID]: "unrestricted",
          [REVIEWS_ID]: "unrestricted",
          [PEOPLE_ID]: "sandboxed",
        }
      },
      "create-queries": {
        PUBLIC: {
          [ORDERS_ID]: "query-builder",
          [PRODUCTS_ID]: "query-builder",
          [REVIEWS_ID]: "query-builder",
        },
      },
    },
  },
});
```

File: e2e/test/scenarios/native-filters/sql-filters-source.cy.spec.js [hunk 2]
```diff
@@ -585,7 +585,7 @@ describeEE("scenarios > filters > sql filters > values source", () => {
     cy.sandboxTable({
       table_id: PRODUCTS_ID,
       attribute_remappings: {
-        attr_uid: ["dimension", ["field", PRODUCTS.ID, null]],
+        attr_cat: ["dimension", ["field", PRODUCTS.CATEGORY, null]],
```
Comment (sloansparger) [inline e2e/test/scenarios/native-filters/sql-filters-source.cy.spec.js:588]: This test was properly broken, as we were sandboxing on the wrong attribute.
It also is testing a filter that aggregates unique values and the request for that data does not include the `is_sandboxed` metadata.

**Review by nemanjaglumac — COMMENTED** [review]
Left some inline comments.

My general sentiment is that we shouldn't be relying on the implementation details (API). I think you did a great job of adding row count assertions and, frankly, those should be more than enough.

We realized we had weak assertions.
The goal was to reinforce them. Whether we do it through the row count, or making sure that some elements are while some others are not on the page is less important.

I vote for more UI assertions and reducing the complexity of the newly introduced helper. Those are just my 2c.

Unfortunately, I cannot make this decision alone so I'll add @ranquild and @kamilmielnik as reviewers.

Thank you for all the hard work you've put into this!

**Comment by sloansparger** [comment]
The original implementation of this PR just focused on asserting the # of results. You provided feedback that we should "Improve assertions by adding negative assertions to the mix (not just asserting on the number of rows - even though this works; try making sure that some value that a user should not see does not render on a page)" ([message](https://metaboat.slack.com/archives/C505ZNNH4/p1724187422396359)).

I agree with this feedback, but like # of rows it has it's own faults since we can't assert _why_ something was filtered, we just hope it's because it's sandboxed. Adding the `is_sandboxed` assertion adds to these guarantees, but ultimately each check is brittle in it's own way (e.g.  `is_sandboxed` is true if only 1 of 10 tables in your query is sandboxed, so fingers crossed you configured the right one!).

As for asserting on the API response, I fully agree this is an anti-pattern generally, but I think there's value in using it here to provide the strongest guarantees we can that our data access and security is working correctly.

I'm will to concede on

**Comment by sloansparger** [comment]
The original implementation of this PR just focused on asserting the # of results. You provided feedback that we should "Improve assertions by adding negative assertions to the mix (not just asserting on the number of rows - even though this works; try making sure that some value that a user should not see does not render on a page)" ([message](https://metaboat.slack.com/archives/C505ZNNH4/p1724187422396359)).

I agree with this feedback, but like # of rows it has it's own faults since we can't assert why something was filtered, we just hope it's because it's sandboxed. Adding the is_sandboxed assertion adds to these guarantees, but ultimately each check is brittle in it's own way (e.g. is_sandboxed is true if only 1 of 10 tables in your query is sandboxed, so fingers crossed you configured the right one!).

As for asserting on the API response, I fully agree this is an anti-pattern generally, but I think there's value in using it here to provide the strongest guarantees we can that our data access and security is working correctly.

File: e2e/test/scenarios/question/notebook-link-to-data-source.cy.spec.ts [hunk 3]
```diff
@@ -524,9 +525,12 @@ describe("scenarios > notebook > link to data source", () => {
           "have.text",
           "Showing 11 rows",
         );
+        assertDatasetReqIsSandboxed(`@modelQuery${ORDERS_MODEL_ID}`);
       });
 
       it("should work for sandboxed users when joined table is sandboxed", () => {
+        cy.intercept("/api/dataset").as("dataset");
+
```
[the remark this answers is not in the document]
Comment (sloansparger) [reply e2e/test/scenarios/question/notebook-link-to-data-source.cy.spec.ts:533]: `dataset` is the default alias used by they sandbox helper.
for some reason that wasn't happening, but i'll double check...

File: e2e/test/scenarios/dashboard-filters/dashboard-filters-source.cy.spec.js [hunk 4]
```diff
@@ -415,6 +416,7 @@ describeEE("scenarios > dashboard > filters", () => {
           cy.signOut();
           cy.signInAsSandboxedUser();
           visitDashboard(card.dashboard_id);
+          assertDatasetReqIsSandboxed(`@dashcardQuery${card.id}`);
```
[the remark this answers is not in the document]
Comment (sloansparger) [reply e2e/test/scenarios/dashboard-filters/dashboard-filters-source.cy.spec.js:419]: Right, this is simply increasing the guarantee that the reason that certain data is present / not present is due to sandboxing.

File: e2e/support/helpers/e2e-permissions-helpers.js [hunk 5]
```diff
@@ -122,3 +122,42 @@ export function assertSameBeforeAndAfterSave(assertionCallback) {
...
+        expect(
+          sandboxedColumnValues.every(val => val === sandboxedFieldAssertion),
+        ).to.equal(
+          true,
+          `Expected every result in column to be equal to: ${sandboxedFieldAssertion}`,
+        );
+      }
+    }
+  });
+}
```
[the remark this answers is not in the document]
Comment (sloansparger) [reply e2e/support/helpers/e2e-permissions-helpers.js:163]: Fair. I have made a comment on this PR to address the boarder conversation. I'll clean this up a bit to see if it makes you feel any better.

[the remark this answers is not in the document]
Comment (sloansparger) [reply e2e/support/helpers/e2e-permissions-helpers.js]: Should be simpler now.

**Review by ranquild — COMMENTED** [review]
I was thinking about this a lot and I believe @nemanjaglumac is correct - we should add negative assertions instead of relying on `is_sandboxed`. We could still check the BE flag but there should be assertions that some data is visible and some other data is not visible at the same time.

File: e2e/test/scenarios/question/notebook-link-to-data-source.cy.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply e2e/test/scenarios/question/notebook-link-to-data-source.cy.spec.ts:535]: I double checked and it doesn't work without the explicit intercept, indeed.

**Review by nemanjaglumac — APPROVED** [review]
Awesome!
Thank you so much, @sloansparger.

🙌

**Pull request merged**